### PR TITLE
Jetpack plugin: always use s0.wp.com and i0.wp.com

### PR DESCRIPTION
**Pull request by sgomes**
For the vast majority of users, connections to `s[0-2].wp.com` and `i[0-2].wp.com` can be established using HTTP/2. When HTTP/2 (and in some cases even HTTP/1.1) is available, it's preferable to establish fewer connections to fewer origins, particularly when the number of resources to be retrieved is small.

As such, and for the vast majority of users, it makes sense to ensure that all connections are made to `s0.wp.com` and `i0.wp.com` rather than using hashes to pick among it and its other siblings.

This PR ensures that the `0` origin is used for every resource, and removes any existing preloading for `1` and `2`.

#### Changes proposed in this Pull Request:
* Always use `s0.wp.com` and `i0.wp.com` instead of picking among it and its other siblings
* Remove preloads for other origins

#### Jetpack product discussion
None.

#### Does this pull request change what data or activity we track or use?
No.

#### Testing instructions:
* Open a test page with comments, while logged into the site
* Ensure that masterbar-related resources load from `s0.wp.com` (e.g. `mustache` and `notes-common-v2`)
* Scan the document and ensure there are no prefetches for `s1.wp.com` or `s2.wp.com`
* Smoke-test masterbar and comment functionality to ensure nothing broke

### CODE REVIEW

**Review by jeherve — COMMENTED** [review]
> it's preferable to establish fewer connections to fewer origins, particularly when the number of resources to be retrieved is small.

Should we then take the opportunity to make a similar change for Photon?
https://github.com/Automattic/jetpack/blob/0e6a3b4d81cc032ba1f55f725c0cd67604ccb16b/projects/plugins/jetpack/functions.photon.php#L156-L178

**Comment by sgomes** [comment]
> Should we then take the opportunity to make a similar change for Photon?

That sounds like a great idea to me!

@josephscott : Any objections to standardising on `i0.wp.com` as well, for Photon requests?

**Comment by josephscott** [comment]
I'm in favor of this change and doing the same for Photon.

**Comment by sgomes** [comment]
Thank you @josephscott and @jeherve ! Switching to "In Progress" while I work on extending this to Photon.

**Comment by kraftbj** [comment]
IIRC, the original rational for multiple subdomains was to push more items to avoid limitations on how many connections a browser would make to a single origin at once. 

Just confirming that this is no longer a concern, or at least less of one. 

With regards to Photon, we sometimes see folks with galleries, for example, have a very large number of images on one page. Any benefit to keeping subdomains?

Overall, just playing a bit of devil's advocate.

**Comment by sgomes** [comment]
> IIRC, the original rational for multiple subdomains was to push more items to avoid limitations on how many connections a browser would make to a single origin at once.
> 
> Just confirming that this is no longer a concern, or at least less of one.
> 
> With regards to Photon, we sometimes see folks with galleries, for example, have a very large number of images on one page. Any benefit to keeping subdomains?
> 
> Overall, just playing a bit of devil's advocate.

Hey @kraftbj!

Things have changed with HTTP/2, which works on different principles. Whereas in HTTP/1.x, a single connection could only be used for a single resource fetch at a time; on HTTP/2, there are multiple streams multiplexed onto a single connection, with each stream corresponding to an individual resource. This essentially shifts scheduling concerns one layer up, allowing for the browser to still handle and prioritise multiple concurrent requests, but doing so on top of a single HTTP-level connection.

There should be no need for multiple connections to multiple origins as long as HTTP-level connections aren't being bandwidth-throttled; but if they are, that's not on our end, and it's very questionable that we should be working around that by connecting to multiple origins at once.

This all came together to effectively change the best practice; whereas before it was a best practice to have multiple origins to allow for more than 4 concurrent resource requests (depending on browser, OS, etc.), this is now no longer a concern because those resource requests are being multiplexed onto a single connection. And as such, the startup costs of establishing multiple connections to multiple origins, particularly for critical path resources (such as many styles and scripts), no longer have any practical upside.

Hope this clears things up, but I'm happy to go into more detail if you're still curious!

**Comment by yogeshbeniwal** [comment]
> With regards to Photon, we sometimes see folks with galleries, for example, have a very large number of images on one page. Any benefit to keeping subdomains?

IMHO Image gallery most of the time are below the fold and are lazy loaded, establishing fewer connections will improve for above the fold content.

+1 for removing extra connections for both.

**Comment by sgomes** [comment]
Rebased on master.

**Comment by sgomes** [comment]
This PR is now ready for review again.

File: projects/plugins/jetpack/functions.photon.php [hunk 1]
```diff
@@ -172,7 +167,7 @@ function jetpack_photon_url( $image_url, $args = array(), $scheme = null ) {
 	 *
 	 * @since 3.4.2
 	 *
-	 * @param string https://i{$subdomain}.wp.com Domain used by Photon. $subdomain is a random number between 0 and 2.
+	 * @param string https://i{$subdomain}.wp.com Domain used by Photon. $subdomain is always 0, currently.
```
Comment (kraftbj) [inline projects/plugins/jetpack/functions.photon.php:170]:
```suggestion
	 * @param string https://i0.wp.com Domain used by Photon.
```

**Review by kraftbj — CHANGES_REQUESTED** [review]
Minor changes to keep things tidy.

**Comment by anomiex** [comment]
> Cool and since we're talking about servers we control, there's no need to worry about HTTP 1.1. Makes sense; thanks.

Well, it also depends on the client's browser being new enough to do HTTP 2. We might want to see if we can see how much traffic uses 2 versus 1.

**Comment by josephscott** [comment]
> Well, it also depends on the client's browser being new enough to do HTTP 2. We might want to see if we can see how much traffic uses 2 versus 1.

This is accurate, but I'll go ahead and argue that I don't think it matters.

Based on https://caniuse.com/http2 data, you'd need to be using at least these versions or older to not have HTTP/2 support:
- IE 10 ( or IE 11 outside of WIndows 10 ), 2012
- Firefox 35, 2015
- Chrome 40, 2015
- Safari 8 ( or Safari 10 depending on OS version ), 2014
- Opera mini or Opera 27 2015
- Android Browser 4.4.4, 2014
- Opera Mobile 12.1, 2012

All of those are definitely old.  The closest one that might be considered even remotely controversial is IE 11 on Windows before 10.  IE 11 was released in 2013, and Windows 10 in 2015.  Given how long ago that was, I don't think those are the conditions that we should be optimizing for.  Especially when combined with the generally low usage of IE 11.

This change has the added benefit of not breaking any of the older versions listed above.  If someone has IE 6 and they want to visit a site, this change doesn't break anything.  It may mean that it is a slower experience than if they were using any version of Chrome/Firefox released in the last three years, but all of the resources would still load and be available.  Here is a case where we can benefit ~99% of users, and the remaining 1% aren't broken, they just don't get the optimal experience.

**Comment by sgomes** [comment]
Thank you, @josephscott , that's a really good summary of the situation!

In general, when a performance concern has a tradeoff, we should strive to make sure that tradeoff aligns with the majority of users. In this situation, the tradeoff (latency vs bandwidth) is currently aligned with the wrong set of users, namely those running really old OSes and browsers. This change shifts things to align with the majority of users (which will only keeping growing even further in the future), without breaking things for anyone as @josephscott points out.

File: projects/plugins/jetpack/functions.photon.php [hunk 2]
```diff
@@ -156,14 +156,9 @@ function jetpack_photon_url( $image_url, $args = array(), $scheme = null ) {
...
-	 * The goal is to have the same subdomain for any particular image to prevent multiple runs resulting in multiple
-	 * images needing to be downloaded by the browser.
-	 *
-	 * We are providing our own generated value by taking the modulus of the crc32 value of the URL.
-	 *
-	 * Valid values are 0, 1, and 2.
+	 * We are now yielding 0 for all requests, since using multiple subdomains is no longer a best practice.
 	 */
-	$subdomain = abs( crc32( $image_host_path ) % 3 );
+	$subdomain = 0;
```
[the remark this answers is not in the document]
Comment (sgomes) [reply projects/plugins/jetpack/functions.photon.php]: Right, I think we can get rid of it entirely 👍

**Review by kraftbj — APPROVED** [review]

**Review by jeherve — APPROVED** [review]
This looks good to me. 

Since Jetpack uses an external library to manage Photon for its tiled gallery block, I think this may need an update as well:
https://github.com/Automattic/wp-calypso/blob/10fe5d2c17197c46cd048560518a1718efb8dc8e/packages/photon/src/index.js#L65

**Comment by sgomes** [comment]
> Since Jetpack uses an external library to manage Photon for its tiled gallery block, I think this may need an update as well: https://github.com/Automattic/wp-calypso/blob/10fe5d2c17197c46cd048560518a1718efb8dc8e/packages/photon/src/index.js#L65

Thank you, @jeherve! It looks like that code already does the right thing. Well, mostly:

```js
function serverFromUrlParts( pathname, isSecure ) {
	if ( isSecure ) {
		return 'i0.wp.com';
	}
	const hash = crc32( pathname );
	const rng = seed( hash );
	const server = 'i' + Math.floor( rng() * 3 );
	debug( 'determined server "%s" to use with "%s"', server, pathname );
	return server + '.wp.com';
}
``` 

It seems to return `i0` for every secure connection, and only picks between all three for images that were already coming from a Photon URL with an insecure `http:` URL. This is likely being done as a flawed proxy for HTTP/2 support, in that HTTP/2 connections will always be secure because of how browser vendors decided to adopt HTTP/2. It's really very strange, though, because it's not checking the user's browser but rather the image URL we're starting from 🤷

I don't think this approach necessarily makes a ton of sense, but it's fine to leave it as-is, since hopefully `http:` Photon URLs are few and far between.

**Comment by sgomes** [comment]
r235975

**Pull request merged**